### PR TITLE
Modify Cholesky derivative

### PR DESCRIPTION
**Pull request by vishwakftw**
The derivative of the Cholesky decomposition was previously a triangular matrix.

Changelog:
- Modify the derivative of Cholesky from a triangular matrix to symmetric matrix

Test Plan:
- Add tests to validate that the derivative of the input to matrix function which uses and doesn't use the Cholesky decomposition are the same.

Ref: #18825

cc: @jacobrgardner

### CODE REVIEW

**Comment by vishwakftw** [comment]
Distributions tests are failing due to MVN log_prob's gradcheck.

**Comment by vishwakftw** [comment]
cc: @neerajprad @fritzo this change affects MVN's `log_prob` function. It would great if you can take a look at this.

**Comment by fritzo** [comment]
@vishwakftw I suspect the test failures are due to the use of `torch.autograd.gradcheck` inside `TestDistributions._gradcheck_log_prob()`, where the input matrix `cov` is constrained to be symmetric (and assumed to be symmetric inside `torch.cholesky`, whereas `gradcheck` is unaware of this constraint. This could be caused if the previous cholesky grad implementation symmetrized its input (e.g. via `cov = 0.5 * (cov + cov.transpose(-1, -2)`) but your new implementation does not.

If my hypothesis is correct, then you could fix in a few different ways:
- symmetrizing inside cholesky grad
- symmetrizing inside `MultivariateNormal` (seems like a lot of overhead)
- forking `gradcheck` to support linear constraints

**Comment by fritzo** [comment]
@vishwakftw I suspect the test failures are due to the use of `torch.autograd.gradcheck` inside `TestDistributions._gradcheck_log_prob()`, where the input matrix `cov` is constrained to be symmetric (and assumed to be symmetric inside `torch.cholesky`, whereas `gradcheck` is unaware of this constraint. This could be caused if the previous cholesky grad implementation symmetrized its input (e.g. via `cov = 0.5 * (cov + cov.transpose(-1, -2)`) but your new implementation does not.

If my hypothesis is correct, then you could fix in a few different ways:
- symmetrizing inside cholesky grad
- symmetrizing inside `MultivariateNormal` (seems like a lot of overhead)
- forking `gradcheck` to support linear constraints

**Comment by SsnL** [comment]
The correct fix, I believe, is to gradcheck `f(x) = (x + x.T).cholesky()`.

File: tools/autograd/templates/Functions.cpp
Comment (SsnL) [inline tools/autograd/templates/Functions.cpp]: could you add some comment here about the ambiguity and why we prefer symmetric grad?

File: test/test_autograd.py
Comment (SsnL) [inline test/test_autograd.py]: I'm not sure if we want to specifically enforce this. Could we just check that the gradient is always symmetric?

File: tools/autograd/templates/Functions.cpp
Comment (vishwakftw) [reply tools/autograd/templates/Functions.cpp]: Sure, I will do that.

File: test/test_autograd.py
Comment (vishwakftw) [reply test/test_autograd.py]: I thought this was an interesting test to have, but I can leave it at the symmetry test. Thank you again.

**Comment by vishwakftw** [comment]
@SsnL @fritzo Are you suggesting something like this with regard to the log_prob test?
```
def mvn_log_prob_gradcheck(mean, cov):
    s = torch.distributions.MultivariateNormal(mean, cov).sample().requires_grad_()
    def gradcheck_func(sample, m, c):
        c = 0.5 * (c + c.transpose(-1, -2))  # symmetrizes the covariance
        return torch.distributions.MultivariateNormal(m, c).log_prob(sample)
    torch.autograd.gradcheck(gradcheck_func, (s, mean, cov), raise_exception=True)
```

**Comment by SsnL** [comment]
Code looks good! Wait for CI to be back and test it.

**Comment by vishwakftw** [comment]
All tests are passing. Could I get a final review on this? Thank you.

File: tools/autograd/templates/Functions.cpp
Comment (SsnL) [inline tools/autograd/templates/Functions.cpp]: I think the root problem isn't about this specific case, but about the assumption made by `cholesky`, i.e., the input being symmetric. It is because of this that the gradient can be freely distributed for each reflected pair of values. The motivation, I think, should be that symmetric gradient leads to (usually) more stable gradient updates, and (if the matrix is updated with SGD-like algorithm) leaves the matrix hermitian. What do you think?

**Review by ssnl — APPROVED** [review]
I have a minor comment on the note. But the patch is great. Thanks!

File: tools/autograd/templates/Functions.cpp
Comment (vishwakftw) [reply tools/autograd/templates/Functions.cpp]: I think this sounds good. I wasn't sure if a motivational statement should be presented here, which is why refrained from adding it.

But yes, I should have elucidated more about the root problem. Please allow me to make changes to the note.

Comment (SsnL) [reply tools/autograd/templates/Functions.cpp]: I think this is where we make a design choice. And a good explanation of the ambiguity and the reason we make the choice will be very useful. :)

Comment (vishwakftw) [reply tools/autograd/templates/Functions.cpp]: I've edited the note, thanks for the suggestions.

**Comment by vishwakftw** [comment]
@SsnL Is this good to go?

**Comment by SsnL** [comment]
Yes, of course. Approval above was supposed to indicate that. Thanks for updating the comment! 

@pytorchbot merge this please

**Pull request closed without merging**
